design/src/rfcs/rfc0036_http_dep_elimination.md

```diff
@@ -0,0 +1,107 @@
+----------------------------------------------
+Customers are impacted in 3 main locations:
+1. HTTP types in Interceptors
+2. HTTP types in `customize(...)`
+3. HTTP types in Connectors
+
+In all three of these cases, users would interact with our `http` wrapper types instead.
+
+
+In the current version of the SDK, we expose public dependencies on the `http` crate in several key places:
+
+1. The `sigv4` crate. The `sigv4` crate currently operates directly on many types from the `http` crate. This is unnecessary and actually makes the crate more difficult to use. Although `http` may be used internally, `http` will be removed from the public API of this crate.
+2. Interceptor Context: `interceptor`s can mutate the HTTP request through an unshielded interface. This requires creating a [wrapper layer around `http::Request`](#http-request-wrapper) and updating already written interceptors.
+3. `aws-config`: `http::Response` and `uri`
+4. A long tail of exposed requests and responses in the runtime crates. Many of these crates will be removed post-orchestrator so this can be temporarily delayed.
+
+<!-- Explain the implementation of this new feature -->
+How to actually implement this RFC
+----------------------------------
+
+### Enabling API evolution
+One key mechanism that we SHOULD use for allowing our APIs to evolve in the future is usage of `~` version bounds for the runtime crates after releasing 1.0.
+
+### Http Request Wrapper
+
+In order to enable HTTP evolution, we will create a set of wrapper structures around `http::Request` and `http::Response`. These will use `http = 0` internally. Since the HTTP crate itself is quite small, including private dependencies on both versions of the crate is a workable solution. In general, we will aim for an API that is close to drop-in compatible to the HTTP crate while ensuring that a different crate could be used as the backing storage.
+
+```rust
+// since it's our type, we can default `SdkBody`
+pub struct Request<B = SdkBody> {
+    // this uses the http = 0.2 request. In the future, we can make an internal enum to allow storing an http = 1
+    http_0: http::Request<B>
+}
+```
+
+**Conversion to/from `http::Request`**
+One key property here is that although converting to/from an `http::Request` **can** be expensive, this is *not* ossification of the API. This is because the API can support converting from/to both `http = 0` and `http = 1` in the future—because it offers mutation of the request via a unified interface, the request would only need to be converted once for dispatch if there was a mismatch (instead of repeatedly). At some point in the future, the `http = 0` representation could be deprecated and removed or feature gated.
+
+**Challenges**
+1. Creating an HTTP API which is forwards compatible, idiomatic and "truthful" without relying on existing types from Hyper—e.g. when adding a header, we need to account for the possibility that a header is invalid.
```

Comment:
Doesn't the current API just panic if its invalid (at least for `HeaderName`)? This could potentially be an area for improvement by not allowing panicking construction methods to be used. Although, that would probably make us lose out on `const fn` constructors.

Reply:
`HeaderName::from_static` panics but the other APIs are fallible